### PR TITLE
[PM-4127] Bugfix - Check original target tab URL before executing deferred action due to reprompt

### PR DESCRIPTION
**Pull request by jprusik**
## Type of change

```
- [X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This revisits a new fix for https://github.com/bitwarden/clients/pull/6280, which introduced a bug that is being reverted to avoid potentially making it to release without a resolution. This PR re-approaches the solve for that issue.

## Code changes

Do a check of the target tab URL to see if it changed during master password reprompt, before executing the form fill.

## Screenshots

Original issue:

https://github.com/bitwarden/clients/assets/1556494/d7c6e4a9-746f-44d4-8a6e-8b8a83c6e880

### CODE REVIEW

File: apps/browser/src/vault/popup/components/vault/view.component.ts [hunk 1]
```diff
@@ -331,11 +331,14 @@ export class ViewComponent extends BaseViewComponent {
...
   private async doAutofill() {
+    const originalTabURL = this.tab.url;
     if (!(await this.promptPassword())) {
       return false;
     }
 
-    if (this.pageDetails == null || this.pageDetails.length === 0) {
+    const tabUrlChanged = !(originalTabURL && this.tab.url && originalTabURL === this.tab.url);
+
+    if (this.pageDetails == null || this.pageDetails.length === 0 || tabUrlChanged) {
```
Comment (jprusik) [inline apps/browser/src/vault/popup/components/vault/view.component.ts:341]: The intent here is to record the target tab URL right before reprompt, and verify it hasn't changed in the time it has taken the user to pass reprompt.

Comment (jprusik) [reply apps/browser/src/vault/popup/components/vault/view.component.ts:341]: One concern of this approach is login pages that update the browser URI at regular intervals for some reason. I expect that's more of an edge case, but curious as to other opinions on the matter.

Comment (cagonzalezcs) [reply apps/browser/src/vault/popup/components/vault/view.component.ts:341]: What if we just tested the base URI without query params in place? 

```
const originalUrlObj = new URL(originalTabURL);
const currentUrlObj = new URL(this.tab.url);

originalTabUrl = `${originalUrlObj.origin}${originalUrlObj.pathname}`;
currentTabUrl = `${currentUrlObj.origin}${currentUrlObj.pathname}`;
```

Comment (cagonzalezcs) [reply apps/browser/src/vault/popup/components/vault/view.component.ts:341]: If the pathname or origin changes, we have issues... doubt that would happen on most login pages though. Query params are less important, if you ask me.

Comment (jprusik) [reply apps/browser/src/vault/popup/components/vault/view.component.ts:341]: If a user is set up with an exact/never/regex for a URI with a specific hash or query param, we'd potentially be missing that (and fill when we shouldn't).

But, even erring on the side of not protecting those (probably extreme) edges is still iteratively better than doing nothing and doesn't preclude us from continuing to revisit this; I think your suggestion is probably the right/pragmatic approach.

**Review by cagonzalezcs — DISMISSED** [review]

File: apps/browser/src/vault/popup/components/vault/view.component.ts [hunk 2]
```diff
@@ -331,11 +331,21 @@ export class ViewComponent extends BaseViewComponent {
...
     }
 
-    if (this.pageDetails == null || this.pageDetails.length === 0) {
+    const currentTabURL = this.tab.url?.length && new URL(this.tab.url);
+
+    const originalTabHostPath =
+      originalTabURL && `${originalTabURL.origin}${originalTabURL.pathname}`;
+    const currentTabHostPath = currentTabURL && `${currentTabURL.origin}${currentTabURL.pathname}`;
+
+    const tabUrlChanged = originalTabHostPath !== currentTabHostPath;
```
[the remark this answers is not in the document]
Comment (jprusik) [reply apps/browser/src/vault/popup/components/vault/view.component.ts:346]: Then we will proceed with our existing logic, as the url has not changed.

**Review by cagonzalezcs — APPROVED** [review]

File: apps/browser/src/vault/popup/components/vault/view.component.ts [hunk 3]
```diff
@@ -331,11 +331,21 @@ export class ViewComponent extends BaseViewComponent {
   }
 
   private async doAutofill() {
+    const originalTabURL = this.tab.url?.length && new URL(this.tab.url);
```
Comment (gbubemismith) [inline apps/browser/src/vault/popup/components/vault/view.component.ts:334]: Not blocking, just a question:  Does it make sense to have some of this implementation live in the `BrowserApi` class cause we have some static functions implemented there?

**Review by gbubemismith — APPROVED** [review]
LGTM!

File: apps/browser/src/vault/popup/components/vault/view.component.ts (hunk 3 above)
Comment (jprusik) [reply apps/browser/src/vault/popup/components/vault/view.component.ts:334]: Yeah, @cagonzalezcs and I have been talking about it (offline and in the linked PRs), but getting that done right is likely going to require some heavy refactors. We're considering this work as an iterative improvement which we can revisit.

**Pull request merged**